### ISSUE TITLE
AerocomEvaluation with EMEP and GHOST colocates data outside of EMEP grid

### ISSUE THREAD

**Issue by ejgal** [metno/pyaerocom#123]
When running the aerocom evaluation with EMEP data and GHOST.EEA.daily there are some matches outside of the EMEP domain, see pictures.

Evaluation interface:
![emep_ghost_web](https://user-images.githubusercontent.com/23613579/83000311-761dc100-a00a-11ea-8b50-865fcb520447.png)

EMEP grid:
![emep_grid](https://user-images.githubusercontent.com/23613579/83000338-7b7b0b80-a00a-11ea-8cd5-2d459342c032.png)

@hansbrenna , @jgliss: Are there some known issues I should be aware of here before looking into this?

**Comment by jgliss** [comment]
Interesting... I have no idea why this is the case... @hansbrenna?

**Comment by jgliss** [comment]
La Reunion is a French Island ;)

**Comment by ejgal** [comment]
Oh, does the colocation routines check for more than matches in lat/lon and time, like country?
I guess this could look like it has matched Plateau Caillou (one of the stations on La Reunion) to some continental polluted region. I'm a bit confused..

![pm10_Plateau_Caillou](https://user-images.githubusercontent.com/23613579/83009687-6789d680-a017-11ea-9683-141325080892.png)

**Comment by jgliss** [comment]
The colocation goes by lat / lon, so please check with EMEPers what is going on there. If these points show up, then I suspect that these coordinates are actually available in the EMEP model data.

**Comment by jgliss** [comment]
Maybe you used a global run of EMEP?

**Comment by ejgal** [comment]
Seems not, after reading in the EMEP data the latitude range is from 30.05 to 81.95, see below.
The units of the EMEP latitudes are units=Unit('degrees_north'), is that fine?
I tried colocating directly using colocate_gridded_ungridded with same results.

Reading the EMEP and GHOST data:
```
from pyaerocom.colocation import colocate_gridded_ungridded
from pyaerocom.io import ReadEMEP
from pyaerocom.io import ReadGhost
rghost = ReadGhost('GHOST.EEA.daily')
ghost = rghost.read('conco3')
filepath = '/lustre/storeB/project/fou/kl/emep/ModelRuns/2019_REPORTING/EMEP01_L20EC_rv4_
33.2018/Base_day.nc'
remep = ReadEMEP(filepath=filepath)
emep = remep.read_var('conco3', ts_type='daily')
```
Printing the latitudes of Plateu Caillou and the EMEP
```
ghost['Plateau Caillou'].latitude                                                                                                                                                                         
emep.latitude
```
gives the following output:
```
-21.009739000000003
DimCoord(array([30.05, 30.15, 30.25, 30.35, 30.45, 30.55, 30.65, 30.75, 30.85,
       30.95, 31.05, 31.15, 31.25, 31.35, 31.45, 31.55, 31.65, 31.75,
       31.85, 31.95, 32.05, 32.15, 32.25, 32.35, 32.45, 32.55, 32.65,
       32.75, 32.85, 32.95, 33.05, 33.15, 33.25, 33.35, 33.45, 33.55,
       33.65, 33.75, 33.85, 33.95, 34.05, 34.15, 34.25, 34.35, 34.45,
       34.55, 34.65, 34.75, 34.85, 34.95, 35.05, 35.15, 35.25, 35.35,
       35.45, 35.55, 35.65, 35.75, 35.85, 35.95, 36.05, 36.15, 36.25,
       36.35, 36.45, 36.55, 36.65, 36.75, 36.85, 36.95, 37.05, 37.15,
       37.25, 37.35, 37.45, 37.55, 37.65, 37.75, 37.85, 37.95, 38.05,
       38.15, 38.25, 38.35, 38.45, 38.55, 38.65, 38.75, 38.85, 38.95,
       39.05, 39.15, 39.25, 39.35, 39.45, 39.55, 39.65, 39.75, 39.85,
       39.95, 40.05, 40.15, 40.25, 40.35, 40.45, 40.55, 40.65, 40.75,
       40.85, 40.95, 41.05, 41.15, 41.25, 41.35, 41.45, 41.55, 41.65,
       41.75, 41.85, 41.95, 42.05, 42.15, 42.25, 42.35, 42.45, 42.55,
       42.65, 42.75, 42.85, 42.95, 43.05, 43.15, 43.25, 43.35, 43.45,
       43.55, 43.65, 43.75, 43.85, 43.95, 44.05, 44.15, 44.25, 44.35,
       44.45, 44.55, 44.65, 44.75, 44.85, 44.95, 45.05, 45.15, 45.25,
       45.35, 45.45, 45.55, 45.65, 45.75, 45.85, 45.95, 46.05, 46.15,
       46.25, 46.35, 46.45, 46.55, 46.65, 46.75, 46.85, 46.95, 47.05,
       47.15, 47.25, 47.35, 47.45, 47.55, 47.65, 47.75, 47.85, 47.95,
       48.05, 48.15, 48.25, 48.35, 48.45, 48.55, 48.65, 48.75, 48.85,
       48.95, 49.05, 49.15, 49.25, 49.35, 49.45, 49.55, 49.65, 49.75,
       49.85, 49.95, 50.05, 50.15, 50.25, 50.35, 50.45, 50.55, 50.65,
       50.75, 50.85, 50.95, 51.05, 51.15, 51.25, 51.35, 51.45, 51.55,
       51.65, 51.75, 51.85, 51.95, 52.05, 52.15, 52.25, 52.35, 52.45,
       52.55, 52.65, 52.75, 52.85, 52.95, 53.05, 53.15, 53.25, 53.35,
       53.45, 53.55, 53.65, 53.75, 53.85, 53.95, 54.05, 54.15, 54.25,
       54.35, 54.45, 54.55, 54.65, 54.75, 54.85, 54.95, 55.05, 55.15,
       55.25, 55.35, 55.45, 55.55, 55.65, 55.75, 55.85, 55.95, 56.05,
       56.15, 56.25, 56.35, 56.45, 56.55, 56.65, 56.75, 56.85, 56.95,
       57.05, 57.15, 57.25, 57.35, 57.45, 57.55, 57.65, 57.75, 57.85,
       57.95, 58.05, 58.15, 58.25, 58.35, 58.45, 58.55, 58.65, 58.75,
       58.85, 58.95, 59.05, 59.15, 59.25, 59.35, 59.45, 59.55, 59.65,
       59.75, 59.85, 59.95, 60.05, 60.15, 60.25, 60.35, 60.45, 60.55,
       60.65, 60.75, 60.85, 60.95, 61.05, 61.15, 61.25, 61.35, 61.45,
       61.55, 61.65, 61.75, 61.85, 61.95, 62.05, 62.15, 62.25, 62.35,
       62.45, 62.55, 62.65, 62.75, 62.85, 62.95, 63.05, 63.15, 63.25,
       63.35, 63.45, 63.55, 63.65, 63.75, 63.85, 63.95, 64.05, 64.15,
       64.25, 64.35, 64.45, 64.55, 64.65, 64.75, 64.85, 64.95, 65.05,
       65.15, 65.25, 65.35, 65.45, 65.55, 65.65, 65.75, 65.85, 65.95,
       66.05, 66.15, 66.25, 66.35, 66.45, 66.55, 66.65, 66.75, 66.85,
       66.95, 67.05, 67.15, 67.25, 67.35, 67.45, 67.55, 67.65, 67.75,
       67.85, 67.95, 68.05, 68.15, 68.25, 68.35, 68.45, 68.55, 68.65,
       68.75, 68.85, 68.95, 69.05, 69.15, 69.25, 69.35, 69.45, 69.55,
       69.65, 69.75, 69.85, 69.95, 70.05, 70.15, 70.25, 70.35, 70.45,
       70.55, 70.65, 70.75, 70.85, 70.95, 71.05, 71.15, 71.25, 71.35,
       71.45, 71.55, 71.65, 71.75, 71.85, 71.95, 72.05, 72.15, 72.25,
       72.35, 72.45, 72.55, 72.65, 72.75, 72.85, 72.95, 73.05, 73.15,
       73.25, 73.35, 73.45, 73.55, 73.65, 73.75, 73.85, 73.95, 74.05,
       74.15, 74.25, 74.35, 74.45, 74.55, 74.65, 74.75, 74.85, 74.95,
       75.05, 75.15, 75.25, 75.35, 75.45, 75.55, 75.65, 75.75, 75.85,
       75.95, 76.05, 76.15, 76.25, 76.35, 76.45, 76.55, 76.65, 76.75,
       76.85, 76.95, 77.05, 77.15, 77.25, 77.35, 77.45, 77.55, 77.65,
       77.75, 77.85, 77.95, 78.05, 78.15, 78.25, 78.35, 78.45, 78.55,
       78.65, 78.75, 78.85, 78.95, 79.05, 79.15, 79.25, 79.35, 79.45,
       79.55, 79.65, 79.75, 79.85, 79.95, 80.05, 80.15, 80.25, 80.35,
       80.45, 80.55, 80.65, 80.75, 80.85, 80.95, 81.05, 81.15, 81.25,
       81.35, 81.45, 81.55, 81.65, 81.75, 81.85, 81.95]),
standard_name='latitude', units=Unit('degrees_north'), long_name='latitude', var_name='lat')
```
Running the colocation with colocate_gridded_ungridded and the colocate data contains the same station.
```
col = colocate_gridded_ungridded(emep, ghost)
col.data.sel(station_name='Plateu Caillou')
```
```
<xarray.DataArray 'conco3' (data_source: 2, time: 365)>
array([[ 27.58721733,  28.27908516,  20.4518013 ,  19.19726753,
         20.61510658,          nan,          nan,          nan,
.......
         30.70896149],
       [ 82.8452301 ,  85.15762329,  84.17182159,  83.16841888,
........
         71.99658966,  72.88525391,  70.90750885,  73.84544373,
         72.87802887]])
Coordinates:
  * data_source   (data_source) object 'GHOST.EEA.daily' None
    var_name      (data_source) <U6 'conco3' 'conco3'
    var_units     (data_source) <U6 'ug m-3' 'ug m-3'
    ts_type_src   (data_source) <U5 'daily' 'daily'
  * time          (time) datetime64[ns] 2018-01-01 2018-01-02 ... 2018-12-31
    station_name  <U84 'Plateau Caillou'
    latitude      float64 -21.01
    longitude     float64 55.27
    altitude      float32 165.0
Attributes:
    data_source:        ['GHOST.EEA.daily', None]
    var_name:           ['conco3', 'conco3']
    ts_type:            daily
    filter_name:        WORLD-noMOUNTAINS
    ts_type_src:        ['daily', 'daily']
    start_str:          20180101
    stop_str:           20181231
    var_units:          ['ug m-3', 'ug m-3']
    vert_scheme:        None
    data_level:         3
    revision_ref:       n/a
    from_files:         ['', 'l', 'u', 's', 't', 'r', 'e', '', 's', 't', 'o',...
    from_files_ref:     None
    stations_ignored:   None
    colocate_time:      False
    obs_is_clim:        False
    pyaerocom:          0.9.0.dev6
    apply_constraints:  None
    min_num_obs:        None
    outliers_removed:   True
    region:             WORLD
    lon_range:          [-180, 180]
    lat_range:          [-90, 90]
    alt_range:          [-1000000.0, 1000.0]
    land_sea:           None
```

**Comment by ejgal** [comment]
Maybe the stations outside of the EMEP grid is matched to the closest point inside the grid?
Is it possible to easily set a custom region with a lat/lon range?

**Comment by hansbrenna** [comment]
This is exactly what happens. 

Adding `latitude=[30,70],longitude=[-25,45] to the `obs_filters` dictionary in `obs_config` fixes it.

**Comment by jgliss** [comment]
This is interesting and is due to the fact that `xarray.DataArray.sel` is used with `method=nearest` in timeseries extraction of `GriddedData`. This does of course not check if any of the input locations is within the lat/lon boundaries of the grid (until recently, pyaerocom was only used for global grids). Should be an easy fix and I will try to provide it today..

**Comment by ejgal** [comment]
I tried running the evaluation with the recent changes, but it couldn't finish and gave the following error. Not sure where in the code this comes from, and running the debugger over the vpn is very slow. Have any of you managed to run spyder on xvis-m4a with X forwarding?

```
Running EMEP / GHOST.EEA.daily (concpm25, concpm25)
Extracting timeseries data from large array (shape: (365, 520, 1200)). This may take a while...
Colocation between model EMEP / concpm25 and obs GHOST.EEA.daily / concpm25 failed: Reason IndexError('Dimensions of indexers mismatch: (slice(None, None, None), <xarray.Variable (latlon: 1523)>\narray([212, 210, 212, ..., 200, 220, 126]), <xarray.Variable (latlon: 1571)>\narray([332, 337, 344, ..., 486, 504, 481]))')
```

**Comment by jgliss** [comment]
@ejgal I never managed to run spyder properly on lustre, but maybe @jgriesfeller knows more.

**Comment by ejgal** [comment]
Looked a bit more into this, see https://github.com/metno/pyaerocom/pull/126/files#r433793518

**Comment by jgliss** [comment]
@ejgal Great if you can fix this and comment on #127. 

You can refer to issues / PRs also in your commit message, in case you did not know, see e.g.:

https://stackoverflow.com/questions/1687262/link-to-the-issue-number-on-github-within-a-commit-message

**Comment by ejgal** [comment]
Ran into a new error after merging the code in #128.

```
Running analysis:
Obs. names: ['GHOST-daily']
Model names: ['EMEP']
Remove outliers: False
Harmonise units: True
Delete existing json files before reanalysis: True
Reanalyse existing colocated NetCDF files: True
Run only colocation (no json files computed): False
Raise exceptions if they occur: True

Running colocation of EMEP against GHOST-daily
PREPARING colocation of EMEP vs. GHOST.EEA.daily
The following variable combinations will be colocated
MODEL-VAR	OBS-VAR
concpm10	concpm10
conco3	conco3
Running EMEP / GHOST.EEA.daily (concpm10, concpm10)
     
Extracting timeseries data from large array (shape: (365, 520, 1200)). This may take a while...
Time series extraction successful. Elapsed time: 4 s
Colocation between model EMEP / concpm10 and obs GHOST.EEA.daily / concpm10 failed: Reason IndexError('list index out of range')
Failed to perform analysis: Traceback (most recent call last):
  File "/home/eirikg/pyaerocom/pyaerocom/colocation_auto.py", line 936, in _run_gridded_ungridded
    resample_how=rshow)
  File "/home/eirikg/pyaerocom/pyaerocom/colocation.py", line 648, in colocate_gridded_ungridded
    grid_stat = grid_stat_data[i]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/eirikg/pyaerocom/pyaerocom/colocation_auto.py", line 380, in run
    self.data[self.model_id] = self._run_gridded_ungridded(var_name)
  File "/home/eirikg/pyaerocom/pyaerocom/colocation_auto.py", line 955, in _run_gridded_ungridded
    raise Exception(msg)
Exception: Colocation between model EMEP / concpm10 and obs GHOST.EEA.daily / concpm10 failed: Reason IndexError('list index out of range')


Deactivating logging in Colocator. Reason: AttributeError("'NoneType' object has no attribute 'write'")
Traceback (most recent call last):
  File "/home/eirikg/pyaerocom/pyaerocom/colocation_auto.py", line 936, in _run_gridded_ungridded
    resample_how=rshow)
  File "/home/eirikg/pyaerocom/pyaerocom/colocation.py", line 648, in colocate_gridded_ungridded
    grid_stat = grid_stat_data[i]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/eirikg/pyaerocom/pyaerocom/colocation_auto.py", line 380, in run
    self.data[self.model_id] = self._run_gridded_ungridded(var_name)
  File "/home/eirikg/pyaerocom/pyaerocom/colocation_auto.py", line 955, in _run_gridded_ungridded
    raise Exception(msg)
Exception: Colocation between model EMEP / concpm10 and obs GHOST.EEA.daily / concpm10 failed: Reason IndexError('list index out of range')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "cfg_ghost_emep.py", line 136, in <module>
    stp.run_evaluation()
  File "/home/eirikg/pyaerocom/pyaerocom/web/aerocom_evaluation.py", line 994, in run_evaluation
    col = self.run_colocation(model_name, obs_name, var_name)
  File "/home/eirikg/pyaerocom/pyaerocom/web/aerocom_evaluation.py", line 777, in run_colocation
    col.run(var_name)
  File "/home/eirikg/pyaerocom/pyaerocom/colocation_auto.py", line 390, in run
    raise Exception(traceback.format_exc())
Exception: Traceback (most recent call last):
  File "/home/eirikg/pyaerocom/pyaerocom/colocation_auto.py", line 936, in _run_gridded_ungridded
    resample_how=rshow)
  File "/home/eirikg/pyaerocom/pyaerocom/colocation.py", line 648, in colocate_gridded_ungridded
    grid_stat = grid_stat_data[i]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/eirikg/pyaerocom/pyaerocom/colocation_auto.py", line 380, in run
    self.data[self.model_id] = self._run_gridded_ungridded(var_name)
  File "/home/eirikg/pyaerocom/pyaerocom/colocation_auto.py", line 955, in _run_gridded_ungridded
    raise Exception(msg)
Exception: Colocation between model EMEP / concpm10 and obs GHOST.EEA.daily / concpm10 failed: Reason IndexError('list index out of range')
```

**Comment by ejgal** [comment]
This is in the loop running over all obs_stat_data, so it seems that `grid_stat_data` are shorter than `obs_stat_data`. Guess this is linked to the changes in #126  or #128, but not sure how.

**Comment by ejgal** [comment]
This seems to be the case. Since we now potentially shorten the length of `grid_stat_data` with `check_domain = True` the lengths of the gridded and ungridded station data might differ.
Since the number of stations in obs_stat_data are determined by `to_station_data_all`in `UngriddedData` we could add lat/lon arguments to the function and only return the stations that are inside this range. This should not add too much overhead since we are already looping over all the stations.

Otherwise we need to fix this inside `colocate_gridded_ungridded` somehow.

@jgliss, any ideas on this?

**Comment by ejgal** [comment]
See [this commit](https://github.com/metno/pyaerocom/commit/d80cdc502b96817b4935252ec3f3be710b2351b8?w=1) for a possible implementation. Ran the colocation again now and the stations outside of the EMEP grid are not included.

**Comment by jgliss** [comment]
@ejgal great, thanks for working on this. Can you add the recent commits to PR #128. And please check the failing tests (I think some tests in test_ungriddeddata.py are failing and also strange that all tests in test_read_ebas.py are skipped...).

**Comment by ejgal** [comment]
Fixed with #128
